File: ruby/estimate_pi.rb
ITERATIONS = 100_000_000

in_sector_hits = 0

ITERATIONS.times.each do
  x, y = rand, rand

  # найдём длину отрезка
  z = Math.sqrt(x ** 2 + y ** 2)

  in_sector_hits += 1 if z <= 1
end

# Вероятность попадания в сектор 1/4 окружности с радиусом 1
p_in_sector = in_sector_hits.to_f / ITERATIONS

# Оценка pi
pi = 4 * p_in_sector

puts "Оценка числа pi = #{pi} после #{ITERATIONS} итераций"

File: go/estimate_pi.go
package main

import (
 "fmt"
 "math"
 "math/rand"
 "time"
)

func main() {
 count := 100_000_000
 inSector := 0

 // Инициализация генератора случайных чисел
 rand.Seed(time.Now().UnixNano())

 for i := 0; i < count; i++ {
  x := rand.Float64()
  y := rand.Float64()
  // Условие с корнем
  if math.Sqrt(x*x+y*y) <= 1.0 {
   inSector++
  }
 }

 pi := 4.0 * float64(inSector) / float64(count)
 fmt.Printf("pi: %f\n", pi)
}
